File: test/eventLoopStats.js
var eventLoopStats = require('..');
var expect = require('chai').expect;

describe('eventLoopStats', function() {
  it('should expose a sense function', function() {
    expect(eventLoopStats.sense).to.be.a('function');
  });

  it('should return stats via the sense function', function(done) {
    setTimeout(function() {
      // provoke execution on the default loop
    }, 500);

    setTimeout(function() {
      var stats = eventLoopStats.sense();
      expect(stats.num).to.be.a('number');
      expect(stats.num).to.be.gt(0);
      expect(stats.sum).to.be.a('number');
      expect(stats.max).to.be.a('number');
      expect(stats.min).to.be.a('number');
      done();
    }, 1000);
  });

  // This pattern will usually allow for at least _two_ on_check calls between
  // the two sense calls. Check the next test for a slightly different pattern.
  it('should detect a blocked event loop', function(done) {
    // Call sense to reset max.
    eventLoopStats.sense();
    // On the next tick, block for 500ms.
    setTimeout(function() {
      var waitUntill = new Date(Date.now() + 500);
      // Block event loop with busy waiting.
      while (waitUntill > new Date()) {}

      // On the next tick, detect stats again - the 500 ms block should have
      // been noticed.
      setTimeout(function() {
        var stats = eventLoopStats.sense();
        expect(stats.max).to.be.gte(490);
        expect(stats.max).to.be.lt(2000);
        expect(stats.sum).to.be.gte(490);
        expect(stats.sum).to.be.lt(2000);

        // At least two on_check calls should have happened (older Node.js versions will call it more often).
        expect(stats.num).to.be.gte(2);

        // Since there are at least two on_check calls, min and max should be different.
        expect(stats.min).to.be.gte(0);
        expect(stats.min).to.be.lt(stats.max);

        done();
      }, 0);
    }, 0);
  });

  // This pattern will typically allow for only _one_ on_check call between the
  // two sense calls (at least on more recent Node.js versions). Check the
  // previous test for a slightly different pattern.
  it('one on_check call should suffice to report correct max duration ', function(done) {
    // Call sense to reset max.
    eventLoopStats.sense();
    var now = Date.now();
    var end = now + 50;
    setTimeout(function() {
      // After 100ms, call sense again - the 50ms block (see below) should have been noticed.
      var stats = eventLoopStats.sense();
      expect(stats.max).to.be.gte(40);
      expect(stats.max).to.be.lt(1000);
      expect(stats.sum).to.be.gte(40);
      expect(stats.sum).to.be.lt(1000);

      // Only one on_check call might have happened.
      expect(stats.num).to.gte(1);

      done();
    }, 100);

    // Directly block for 50ms right in this tick, synchronously.
    while (Date.now() < end) {}
  });
});

File: src/eventLoopStats.cc
#include <nan.h>

using namespace v8;

// Casting -1 to an uint will give the max uint value
const uint32_t maxPossibleUint32 = -1;
const uint64_t maxPossibleUint64 = -1;
const uint32_t minPossibleUint32 = 0;

uv_check_t check_handle;

// The minimum event loop duration since the last sense() call.
uint32_t min;
// The maximum event loop duration since the last sense() call.
uint32_t max;
// The sum of event loop durations the last sense() call.
uint32_t sum;
// The number of event loop iterations since the last sense() call.
uint32_t num;

uint64_t previous_now = maxPossibleUint64;

// This will be called after each sense call.
void reset() {
  min = maxPossibleUint32;
  max = minPossibleUint32;
  sum = 0;
  num = 0;
}

// This will be called once per event loop iteration, right after polling
// for i/o. See http://docs.libuv.org/en/v1.x/design.html#the-i-o-loop for a
// general overview and http://docs.libuv.org/en/v1.x/check.html for details.
void on_check(uv_check_t* handle) {
  // An earlier incarnation of event-loop-stats checked uv_hrtime() against
  // uv_now(handle->loop), which doesn't seem to work -- for example, it does
  // not detect long event loop iterations caused by a synchronous block due to
  // long running for- or while-loops. Checking against the time of the same
  // point at the last iteration of the event loop also covers these cases.

  const uint64_t now = uv_hrtime();
  uint32_t duration;

  if (previous_now >= now) {
    // This only happens on the very first call on_check call. Since we have no
    // timestamp to compare to from an earlier on_check call, we start by
    // assuming an event loop lag of zero.
    duration = 0;
  } else {
    // Calculate the duration since the last on_check call - this is the
    // event loop lag.
    // And convert to milliseconds (uv_hrtime yields nanos).
    duration = (now - previous_now) / static_cast<uint64_t>(1e6);
  }

  // save min/max values
  if (duration < min) {
    min = duration;
  }
  if (duration > max) {
    max = duration;
  }

  // Sum up all durations between two consecutive sense() calls.
  sum += duration;

  // Simply count all event loop iterations between two sense() calls.
  num += 1;

  // Save the current timestamp for the next on_check call for comparison.
  previous_now = now;
}


static NAN_METHOD(sense) {
  // Reset min and max counters when there were no calls.
  if (num == 0) {
    min = 0;
    max = 0;
  }

  Local<Object> obj = Nan::New<Object>();
  Nan::Set(
    obj,
    Nan::New("min").ToLocalChecked(),
    Nan::New<Number>(static_cast<double>(min))
  );
  Nan::Set(
    obj,
    Nan::New("max").ToLocalChecked(),
    Nan::New<Number>(static_cast<double>(max))
  );
  Nan::Set(
    obj,
    Nan::New("num").ToLocalChecked(),
    Nan::New<Number>(static_cast<double>(num))
  );
  Nan::Set(
    obj,
    Nan::New("sum").ToLocalChecked(),
    Nan::New<Number>(static_cast<double>(sum))
  );

  reset();

  info.GetReturnValue().Set(obj);
}


NAN_MODULE_INIT(init) {
  reset();

  uv_check_init(uv_default_loop(), &check_handle);
  uv_check_start(&check_handle, reinterpret_cast<uv_check_cb>(on_check));
  uv_unref(reinterpret_cast<uv_handle_t*>(&check_handle));

  Nan::Set(target,
    Nan::New("sense").ToLocalChecked(),
    Nan::GetFunction(Nan::New<FunctionTemplate>(sense)).ToLocalChecked()
  );
}

NODE_MODULE(eventLoopStats, init)
